### ISSUE TITLE
Readme?

### ISSUE THREAD

**Issue by swordsx48** [corpnewt/USBMap#21]
I'm so confused. I have no idea how to use USBmap itself. Every link I've seen says to read the readme. But there is no info on how to actually use the application. I hit D and then what? I'm so confused...how does everyone just know what to do? I must be trippin...

**Comment by corpnewt** [comment]
I've been pretty busy lately, so I haven't had the time/energy to flesh out the readme in the repo.  I could likely answer some quick questions here though in the meantime.

-CorpNewt

**Comment by BooBerry** [comment]
I believe what's being asked is something like a quick start guide, like quick instructions/steps on what to do - basically the TL;DR version of mapping your USB.

**Comment by swordsx48** [comment]
> I believe what's being asked is something like a quick start guide, like quick instructions/steps on what to do - basically the TL;DR version of mapping your USB.

Yes sir! From what I'm gathering from YouTube videos, the Readme did contain a quick start guide / steps - it seems it was removed over time? I think the interface of the script changed too so that's understandable.

**Comment by rbugajewski** [comment]
I’m as confused as @swordsx48 and came here with the same conclusions. Almost everything points at this README, and it is very nice regarding the theoretical background, but I still don’t know how to properly use the tool.

1. I let the tool detect, then I selected all ports and added the USBMap to my EFI.
2. I have three controllers on my board so I don’t go over the port limit, and there seems to be no rename necessary.
3. I don’t know if this is everything what was needed, or if I had to plug in devices to every port at some point so that everything works correctly.
4. And I don’t even know how to properly verify the map.

**Comment by rbugajewski** [comment]
Another question: the guide has a fat banner at the top that Intel USB mapping is outdated, and it points to this README, but can’t I just use this guide to do the mapping on an AMD OpenCore build?

**Comment by Oxford999** [comment]
I don't know how to proceed either. I tried to follow the Dortania guide but it says the USB mapping section is out of date and points to the readme for this project. 

I tried using USBMap without any guidance (other than its interface) and I don't know if the results are correct, particularly since they are different from the results given to me by someone claiming they created a working configuration with it for my machine. There are substantial differences in the info.plist files (between his, apparently made with USBMap on April 9th -- and the one I made with USBMap). I didn't know if I have to do something to correct the listing for the ports detected by USBMap because they all say USB 3, even though the machine only has two USB 3 ports.

I followed the  _varszegimarcell / Optiplex-3020-Hackintosh-OpenCore_  guide with the appropriate hardware but sleep is broken and the guide didn't have anything in it about USBMap because its original creator was only using USBInjectAll. I am going to try changing to the April 9th USBMap kext and info.plist someone else gave me to see if that will help but the differences between that info.plist and what I got from the current version of USBMap make me wonder if there are bugs that were fixed since April 9th in USBMap that are particularly important to have fixed for the machine I'm trying to get working.

**Comment by ak1de** [comment]
I am struggling with the very same problem, a proper description on how to use USBMap is missing.

I did the detection step. I also identified the USB ports I want to use going forward, leveraging an USB stick.
I selected/highlighted all those and made sure I don't exceed the max. of 15. But what's next?
Creating an USBMap.kext by pressing 'K'?
I did that and looked into the Info.plist of the created USBMap.kext package. It seemed to me that all detected USB ports are listed there, not just those I wanted to map.
So, what I am supposed to do?

**Comment by corpnewt** [comment]
> It seemed to me that all detected USB ports are listed there, not just those I wanted to map.

The current functionality when creating the USBMap.kext is to list all ports, but change the `port` key under any disabled ports to `#port`.  The OS will ignore those - but you still retain the entry in case you would like to adjust the map later.

Ultimately it sounds like you mapped properly to me.

-CorpNewt

**Comment by ak1de** [comment]
> > It seemed to me that all detected USB ports are listed there, not just those I wanted to map.
> 
> The current functionality when creating the USBMap.kext is to list all ports, but change the `port` key under any disabled ports to `#port`. The OS will ignore those - but you still retain the entry in case you would like to adjust the map later.
> 
> Ultimately it sounds like you mapped properly to me.
> 
> -CorpNewt

Thanks for this quick reply. That was helpful!

**Comment by HEPBO3AH** [comment]
Please write a more detailed instruction. 
This is very hard to follow due to lack of details. For example between 5 and  6. Do I exit the Discover ports mode?
Nevermind the crashing like I have which is most likely due to python versions or some other dependency missing. :(

**Comment by grodid0** [comment]
Hello,

 I am puzzled by the Readme of this USBMap tool, cause I want to create an USBMap.kext file.

The Quick Start instructions says

![image](https://user-images.githubusercontent.com/121382054/209466211-0eb5a5ed-3b31-4d18-ac7d-e83995b4a351.png)

So far so good.

But it seems to me that points 4+5 and 7+8 repeat the same actions. Is that correct ? Or is this a mistyping ?
If not, does that me that one has to plug and unplug **twice** all the ports with both types of USB sticks ?
Sorry if this questions seem stupid.

Many thanks, Regards, G.

**Comment by swordsx48** [comment]
Hey there @grodid0 !
I've got it figured out, I'll lay it out for everyone here.

1. Go to USBtoolbox, and use that instead, which has Windows and mac implementations.
2. In that, you also plug a 2.0 and 3.0 device into each port, preferably from a windows boot, whether that is a usb PE or an actual install.
3. You toggle on and off the ports you want, sticking under the 15 port limit (probably at 14 because the way ports are paired)
4. once you're done, be sure to put the usbtoolbox.kext and your custom utbmap.kext

works like a charm!!

**Comment by grodid0** [comment]
Hey,

 Thanks a lot for your reply.

 But, when going to use masOS.py inside of the macOS.zip package for USBtoolbox,
I do not find the command file.
What must I type to run the tool, please ?

I feel dumb…

Thanks,  Regards,  G.

**Comment by swordsx48** [comment]
I strongly recommend using it from a windows PE (live USB) or a windows installation

**Comment by grodid0** [comment]
Hey,

 I understand your advise.

 However, I was absolutely unable to find a convenient Windows thing to use USBToolBox.

 I am sorry to ask you once more, to explain to me precisely how to use the macOS implementation
for this stuff. I know about python, but I don’t see in this case how to launch the proper script/command.

 Many thanks,   Regards,  G.

**Comment by swordsx48** [comment]
What do you mean by "convenient Windows thing"?
If you don't use Windows, it will be a lot harder. 
try this guide https://www.reddit.com/r/hackintosh/comments/ta1ef4/guide_easy_usb_mapping_with_usbtoolbox_on_windows/
install windows on an external ssd if you have to - though you can just run a windows installer and run usbtoolbox from the command prompt as well

**Comment by grodid0** [comment]
Hey,

 Right now, I was able to run masOS.py on my Ventura setup.
And I was able to produce the UTBMap.kext provided in attachment.

￼

But I am not sure this kext is correct.

￼

 The display produced after the K step (build kext) of the tool-master/macOS.py
Script is the following :

￼

 What should I do then ?

 What means « Select ports » ? What for ?

 What means « Enable/Disable ports » ? What for ?

 I tried to read this help site to no avail :

https://chriswayg.gitbook.io/opencore-visual-beginners-guide/alternatives/usb-mapping-on-windows￼
USB Mapping on Windows
chriswayg.gitbook.io

For example, should I plug into each port both one USB 2 and USB 3 stick, or not ?

Also, from this map above, which ports should I toggle with « type 0 »,  and with one with «  type 3 » ?

 What do you think ?

 Many thanks,   Regards,   G.
